cmd/finch/version.go

```diff
@@ -4,27 +4,117 @@
 package main
 
 import (
+	"encoding/json"
 	"fmt"
+	"io"
 
 	"github.com/spf13/cobra"
 
+	"github.com/runfinch/finch/pkg/command"
+	"github.com/runfinch/finch/pkg/flog"
+	"github.com/runfinch/finch/pkg/lima"
 	"github.com/runfinch/finch/pkg/version"
 )
 
-var finchVersion = version.Version
-
-func newVersionCommand() *cobra.Command {
+func newVersionCommand(limaCmdCreator command.LimaCmdCreator, logger flog.Logger, stdOut io.Writer) *cobra.Command {
 	versionCommand := &cobra.Command{
 		Use:   "version",
 		Args:  cobra.NoArgs,
 		Short: "Shows Finch version information",
-		RunE:  versionAction,
+		RunE:  newVersionAction(limaCmdCreator, logger, stdOut).runAdapter,
 	}
 
 	return versionCommand
 }
 
-func versionAction(cmd *cobra.Command, args []string) error {
-	fmt.Println("Finch version:", finchVersion)
+type versionAction struct {
+	creator command.LimaCmdCreator
+	logger  flog.Logger
+	stdOut  io.Writer
+}
+
+// NerdctlVersionOutput captures the nerdctl version.
+type NerdctlVersionOutput struct {
+	Client NerdctlClientOuput  `json:"Client"`
+	Server NerdctlServerOutput `json:"Server"`
+}
+
+// NerdctlClientOuput captures the nerdctl Client output.
+type NerdctlClientOuput struct {
+	Version    string                    `json:"Version"`
+	GitCommit  string                    `json:"GitCommit"`
+	GoVersion  string                    `json:"GoVersion"`
+	Os         string                    `json:"Os"`
+	Arch       string                    `json:"Arch"`
+	Components []NerdctlComponentsOutput `json:"Components"`
+}
+
+// NerdctlServerOutput captures the nerdctl Server output.
+type NerdctlServerOutput struct {
+	Components []NerdctlComponentsOutput `json:"Components"`
+}
+
+// NerdctlComponentsOutput captures the nerdctl components output.
+type NerdctlComponentsOutput struct {
+	Name    string `json:"Name"`
+	Version string `json:"Version"`
+	Details struct {
+		GitCommit string `json:"GitCommit"`
+	}
+}
+
+func newVersionAction(creator command.LimaCmdCreator, logger flog.Logger, stdOut io.Writer) *versionAction {
+	return &versionAction{creator: creator, logger: logger, stdOut: stdOut}
+}
+
+func (va *versionAction) runAdapter(cmd *cobra.Command, args []string) error {
+	return va.run()
+}
+
+func (va *versionAction) run() error {
+	status, err := lima.GetVMStatus(va.creator, va.logger, limaInstanceName)
+	//nolint: gocritic // will change to switch statement later
+	if err != nil {
+		fmt.Fprintf(va.stdOut, "Finch version:\t%s", version.Version)
+		return err
+	} else if status != lima.Running {
+		fmt.Fprintf(va.stdOut, "Finch version:\t%s", version.Version)
+	} else {
```

Comment:
- We can use early return here to reduce the nesting levels of the code below.
- We can create a wrapper function that will call `fmt.Fprintf(va.stdOut, "Finch version:\t%s", version.Version)` if the returned error is not nil so that we can remove the duplicated statements (there're 2 more in the code below).
- Return an error when the VM is not running: Checked nerdctl's code, and it also returns an error if some steps fail (after printing whatever is available).

```suggestion
	if err != nil {
		return fmt.Error("failed to get VM status: %w", err)
	}
	if status != lima.Running {
		return errors.New("detailed version info is unavailable because VM is not running")
	}
```

Reply:
My bad, I meant the following:

```go
func (va *versionAction) run() error {
	if err := va.printVersion(); err != nil {
		// Now we only need this statement once, which is clearer and easier to maintain.
		fmt.Fprintf(va.stdOut, "Finch version:\t%s\n", version.Version)
		return err
	}
	return nil
}

func (va *versionAction) printVersion() error {
	// Move the main logic to this function
	// ...
}
```

Reply:
Please create a follow-up PR for this. Resolving this now.

Reply:
https://github.com/runfinch/finch/pull/147